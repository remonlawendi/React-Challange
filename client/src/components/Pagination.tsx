import React, { Children, FC, Fragment } from "react";
import Styled from "styled-components";

const StyledPageLink = Styled.li`
border: 1px solid #CCC;
color: #CCC;
box-shadow: 1px 1px 5px -5px black;
border-radius: 100%;
height: 30px;
width: 30px;
font-size: 1.1em;
display: flex;
justify-content: center;
align-items: center;
padding: 1px;
margin: 3px;
cursor: pointer;
background-color: ${(props: any) => props.active ? "#5652BF" : "auto"};
&:hover {
  background-color: #5652BF;
  color: white;
  border-color: transparent;
  box-shadow: 1px 1px 8px -5px #6662CF;
}
`;
const PageLink = (props: any) => <StyledPageLink {...props} />;

interface IPagination {
    className?: string;
    pageCount: number;
    currentPage: number;
    setCurrentPage: (pageNumber: number) => void;
}

const Pagination: FC <IPagination> = ({ className, pageCount, currentPage, setCurrentPage }) => {
const hasPrev = currentPage > 0;
const hasNext = currentPage + 1 < pageCount;
return <ul className={className}>
  { pageCount ?
    <Fragment>
      <PageLink onClick={() => setCurrentPage(0)}>{"<<"}</PageLink>
      <PageLink
        disabled={!hasPrev}
        onClick={() => hasPrev && setCurrentPage(currentPage - 1)}
      >
        {"<"}
      </PageLink>
      {[...Array(pageCount)].map((p, i) =>
        <PageLink key={`page-${i}`} index={i} active={i === currentPage} onClick={() => setCurrentPage(i)}>
          {i + 1}
        </PageLink>)
      }
      <PageLink disabled={!hasNext} onClick={() => hasNext && setCurrentPage(currentPage + 1)}>
        {">"}
      </PageLink>
      <PageLink onClick={() => setCurrentPage(pageCount - 1)}>
        {">>"}
      </PageLink>
    </Fragment> : ""
  }
</ul>;
};

export default Styled(Pagination)`
  display: flex;
  justify-content: center;
  width: 100%;
`;
