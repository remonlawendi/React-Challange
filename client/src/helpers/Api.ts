import Axios, { AxiosInstance, AxiosRequestConfig, AxiosResponse, CancelTokenSource } from "axios";
import ICase from "../models/ICase";
import IFeature from "../models/IFeature";

const config: AxiosRequestConfig = {
    baseURL: "https://bikewise.org:443/api/v2/",
    responseType: "json",
};

const axios: AxiosInstance = Axios.create(config);
export const fetchSourceAll: CancelTokenSource = Axios.CancelToken.source();
export const fetchSourceDetails: CancelTokenSource = Axios.CancelToken.source();


interface ICasesResponse {
    incidents: ICase[];
}

interface ApiResponse<T> extends AxiosResponse<T> {
    data: T;
    status: number;
    statusText: string;
    headers: any;
    config: AxiosRequestConfig;
}
/**
 * Description Get all theft cases from Bikewise API, it's not paginated due to API limitations
 */
export const fetchAllCases = async (search?: string, from?: string, to?: string) => {
   const toTimeStamp = (date?: string) => date ? Math.ceil(Date.parse(date) / 1000) : "";
   const theftCases: ApiResponse<ICasesResponse> = await axios.request<ICasesResponse>({
     url: "/incidents",
     method: "get",
     params: { proximity: "berlin", incident_type: "theft", query: search, occurred_before: toTimeStamp(to), occurred_after: toTimeStamp(from) },
     cancelToken: fetchSourceAll.token,
   });
   return theftCases.data.incidents;
};
interface ICaseResponse {
    incident: ICase;
}
interface ILocationsResponse {
     features: IFeature[];
}

/**
 * Description get and return theft case details
 */
export const fetchCaseDetails = async (id: string) => {
  const theftCase: ApiResponse<ICaseResponse> = await axios.request<ICaseResponse>({
     url: `/incidents/${id}`,
     method: "get",
     params: { proximity: "berlin", incident_type: "theft" },
     cancelToken: fetchSourceDetails.token,
   });
  const { incident } = theftCase.data;
  const features: ApiResponse<ILocationsResponse> = await axios.request<ILocationsResponse>({
     url: "/locations",
     method: "get",
     params: { query: incident.title, proximity: "berlin", incident_type: "theft" },
     cancelToken: fetchSourceDetails.token,
   });
  incident.feature = features.data.features.find((f) => f.properties.id === incident.id);
  return incident;
};
