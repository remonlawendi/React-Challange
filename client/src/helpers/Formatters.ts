export const FormatUTCDate = (UTC: number): string => {
    const date = new Date(UTC * 1000);
    return date.toDateString();
};
