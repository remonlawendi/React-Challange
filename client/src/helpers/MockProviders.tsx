
import React from "react";
import { Provider } from "react-redux";
import { Router } from "react-router-dom";
import { createStore } from "redux";
import history from "../History";
import { reducer } from "../reducers";

const store: any = createStore(reducer);

export const withMockProviders = (Component: any) =>
<Router history={history}>
  <Provider store={store}>{Component}</Provider>
</Router>;
