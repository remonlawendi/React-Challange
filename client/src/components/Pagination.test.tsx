import { fireEvent, render } from "@testing-library/react";
import React from "react";
import { withMockProviders } from "../helpers/MockProviders";
import Pagination from "./Pagination";

describe("Pagniate component", () => {
    // Arrange

    const setCurrentPage = jest.fn();

    // Act
    const { getByText } = render(
      withMockProviders(<Pagination pageCount={10} currentPage={1} setCurrentPage={setCurrentPage} />),
    );

    // Assert
    it("should contain middle element range", () => {
       getByText("5");
    });
    it("should call setCurrentPage on Click", () => {
      fireEvent.click(getByText("1"));
      expect(setCurrentPage.mock.calls).toMatchSnapshot();
    });
});
