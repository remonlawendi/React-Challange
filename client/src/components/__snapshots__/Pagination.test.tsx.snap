// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagniate component should call setCurrentPage on Click 1`] = `
Array [
  Array [
    0,
  ],
]
`;
